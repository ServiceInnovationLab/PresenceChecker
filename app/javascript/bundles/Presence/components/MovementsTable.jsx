import React from 'react';
import PropTypes from 'prop-types';
import { format } from 'date-fns';

import { firstCharCap } from '../../../utilities/utilities';

import Table from '../../../components/Table';

import { legacyParse, convertTokens } from "@date-fns/upgrade/v2";

export default class MovementsTable extends React.Component {
  static propTypes = {
    movements: PropTypes.array,
  };

  isIndefiniteVisa = movement => {
    if(movement !== 'null') {
      let indefinite_visas = ['P', 'A', 'R'];
      return indefinite_visas.indexOf(movement) === -1 ? 'movements-table-highlight--negative' : 'movements-table-highlight--positive';
    }
  }

  render() {
    const { movements } = this.props;

    return (
      <section className="movements-table-area">
        <header className="movements-table-header">
          <h2>Movements</h2>
        </header>

        <div className="movements-table-wrapper">
          <Table className="presence-table movements-table">
            {movements.map((item, index) => (
              <tr key={`movement_${index}`} className={this.isIndefiniteVisa(item.visa_type)}>
                <td className="header-container has-bottom-border has-right-border">
                  <h3 className="is-dark">{firstCharCap(item.direction)}</h3>
                </td>
                <td className="has-bottom-border has-right-border">
                  {format(legacyParse(item.carrier_date_time), convertTokens('dd mm yyyy'))}
                </td>
                <td className="has-bottom-border">
                  {item.visa_type}
                </td>
              </tr>
            ))}
          </Table>
        </div>
      </section>
    );
  }
}
