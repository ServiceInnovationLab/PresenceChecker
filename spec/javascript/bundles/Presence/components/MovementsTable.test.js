import React from 'react';
import { configure, mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';

import MovementsTable from '../../../../../app/javascript/bundles/Presence/components/MovementsTable';

configure({ adapter: new Adapter() });

describe('<MovementsTable />', () => {
  let props = {
    movements: [
      {
        carrier_date_time: '2018-02-12T00:00:00.000Z',
        created_at: '2019-03-03T22:55:25.179Z',
        direction: 'arrival',
        id: 93,
        identity_id: 13,
        updated_at: '2019-03-03T22:55:25.179Z',
        visa_type: 'R'
      },
      {
        carrier_date_time: '2018-01-20T00:00:00.000Z',
        created_at: '2019-03-03T22:55:25.177Z',
        direction: 'arrival',
        id: 91,
        identity_id: 13,
        updated_at: '2019-03-03T22:55:25.177Z',
        visa_type: 'X'
      },
      {
        carrier_date_time: '2018-01-20T00:00:00.000Z',
        created_at: '2019-03-03T22:55:25.177Z',
        direction: 'departure',
        id: 92,
        identity_id: 13,
        updated_at: '2019-03-03T22:55:25.177Z',
        visa_type: null
      }
    ],
  };

  it('renders without error', () => {
    mount(<MovementsTable {...props} />);
  });

  it('renders correctly without failing', () => {
    const tree = renderer
      .create(<MovementsTable {...props} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  /* The table doesn't have any headers so we're testing that there are 3 columns and the
     third column should be the visa type column */
  it('should have a visa type column', () => {
    const wrapper = mount(<MovementsTable {...props} />);
    expect(wrapper.find('.movements-table tr').length).toEqual(3);
  });

  it('should highlight the table if it has an indefinite visa', () => {
    const wrapper = mount(<MovementsTable {...props} />);
    expect(wrapper.find('.movements-table .movements-table-highlight--positive td').last().text()).toBe('R');
  });

  it('should not highlight the table if it does not have an indefinite visa', () => {
    const wrapper = mount(<MovementsTable {...props} />);
    expect(wrapper.find('.movements-table .movements-table-highlight--negative td').last().text()).toBe('');
  });

  it('should not highlight a row at all if the visa type is nil', () => {
    const wrapper = shallow(<MovementsTable {...props} />);
    expect(wrapper.find('.movements-table').children().last().contains(<td
      className="has-bottom-border"
    />)).toEqual(true);
    expect(wrapper.find('.movements-table').children().last().hasClass('movements-table-highlight--positive')).toEqual(false);
  });

});
