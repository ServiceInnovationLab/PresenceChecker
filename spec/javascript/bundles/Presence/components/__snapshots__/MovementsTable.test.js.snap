// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovementsTable /> renders correctly without failing 1`] = `
<section
  className="movements-table-area"
>
  <header
    className="movements-table-header"
  >
    <h2>
      Movements
    </h2>
  </header>
  <div
    className="movements-table-wrapper"
  >
    <table
      className="presence-table movements-table"
    >
      <tbody>
        <tr
          className="movements-table-highlight--positive"
        >
          <td
            className="header-container has-bottom-border has-right-border"
          >
            <h3
              className="is-dark"
            >
              Arrival
            </h3>
          </td>
          <td
            className="has-bottom-border has-right-border"
          >
            12 Feb 2018
          </td>
          <td
            className="has-bottom-border"
          >
            R
          </td>
        </tr>
        <tr
          className="movements-table-highlight--negative"
        >
          <td
            className="header-container has-bottom-border has-right-border"
          >
            <h3
              className="is-dark"
            >
              Departure
            </h3>
          </td>
          <td
            className="has-bottom-border has-right-border"
          >
            20 Jan 2018
          </td>
          <td
            className="has-bottom-border"
          />
        </tr>
      </tbody>
    </table>
  </div>
</section>
`;
